File: src/editor/input.zig
const std = @import("std");
const terminal = @import("../terminal.zig");
const Editor = @import("Editor.zig");

fn ctrlKey(c: u8) u8 {
    return c & 0x1f;
}

/// Use poll() to check if stdin has data available within `timeout_ms`.
fn stdinHasData(timeout_ms: i32) bool {
    const POLLIN = 0x0001;
    var pfd = [1]std.posix.pollfd{.{
        .fd = std.posix.STDIN_FILENO,
        .events = POLLIN,
        .revents = 0,
    }};
    const n = std.posix.poll(&pfd, timeout_ms) catch return false;
    return n > 0 and (pfd[0].revents & POLLIN) != 0;
}

/// Drain any stale bytes sitting in stdin (e.g. from terminal responses).
pub fn drainInput() void {
    while (stdinHasData(0)) {
        var discard: [64]u8 = undefined;
        _ = std.posix.read(std.posix.STDIN_FILENO, &discard) catch break;
    }
}

fn readByte() !?u8 {
    var buf: [1]u8 = undefined;
    const n = try std.posix.read(std.posix.STDIN_FILENO, &buf);
    if (n == 0) return null;
    return buf[0];
}

pub fn readKey() !Editor.Key {
    // Block until a byte is available
    var buf: [1]u8 = undefined;
    while (true) {
        const n = try std.posix.read(std.posix.STDIN_FILENO, &buf);
        if (n == 0) continue;
        break;
    }
    const c = buf[0];

    if (c == 127) return Editor.Key{ .char = 127 }; // backspace
    if (c != '\x1b') return Editor.Key{ .char = c };

    // Got ESC byte — use poll to check if more bytes follow (escape sequence).
    // A bare ESC keypress won't have follow-up bytes.
    if (!stdinHasData(50)) return Editor.Key{ .char = '\x1b' };

    const s0 = (try readByte()) orelse return Editor.Key{ .char = '\x1b' };
    if (!stdinHasData(50)) return Editor.Key{ .char = '\x1b' };
    const s1 = (try readByte()) orelse return Editor.Key{ .char = '\x1b' };

    if (s0 == '[') {
        if (s1 >= '0' and s1 <= '9') {
            if (!stdinHasData(50)) return Editor.Key{ .char = '\x1b' };
            const s2 = (try readByte()) orelse return Editor.Key{ .char = '\x1b' };
            if (s2 == '~') {
                return switch (s1) {
                    '1' => Editor.Key.home,
                    '3' => Editor.Key.delete,
                    '4' => Editor.Key.end,
                    '5' => Editor.Key.page_up,
                    '6' => Editor.Key.page_down,
                    '7' => Editor.Key.home,
                    '8' => Editor.Key.end,
                    else => Editor.Key{ .char = '\x1b' },
                };
            }
            // Check for shift+arrow: ESC [ 1 ; 2 <dir>
            if (s1 == '1' and s2 == ';') {
                if (!stdinHasData(50)) return Editor.Key{ .char = '\x1b' };
                const mod = (try readByte()) orelse return Editor.Key{ .char = '\x1b' };
                if (!stdinHasData(50)) return Editor.Key{ .char = '\x1b' };
                const dir = (try readByte()) orelse return Editor.Key{ .char = '\x1b' };
                if (mod == '2') {
                    return switch (dir) {
                        'A' => Editor.Key.shift_arrow_up,
                        'B' => Editor.Key.shift_arrow_down,
                        'C' => Editor.Key.shift_arrow_right,
                        'D' => Editor.Key.shift_arrow_left,
                        else => Editor.Key{ .char = '\x1b' },
                    };
                }
            }
        }
        return switch (s1) {
            'A' => Editor.Key.arrow_up,
            'B' => Editor.Key.arrow_down,
            'C' => Editor.Key.arrow_right,
            'D' => Editor.Key.arrow_left,
            'H' => Editor.Key.home,
            'F' => Editor.Key.end,
            else => Editor.Key{ .char = '\x1b' },
        };
    } else if (s0 == 'O') {
        return switch (s1) {
            'H' => Editor.Key.home,
            'F' => Editor.Key.end,
            else => Editor.Key{ .char = '\x1b' },
        };
    }

    return Editor.Key{ .char = '\x1b' };
}

/// Returns false to signal quit.
pub fn processKeypress(self: *Editor) !bool {
    const key = try readKey();

    switch (key) {
        .char => |c| {
            if (c == ctrlKey('q')) {
                if (self.dirty > 0 and self.quit_times > 0) {
                    self.setStatusMessage("WARNING: Unsaved changes. Press Ctrl-Q {d} more time(s) to quit.", .{self.quit_times});
                    self.quit_times -= 1;
                    return true;
                }
                return false;
            } else if (c == ctrlKey('s')) {
                try self.save();
            } else if (c == ctrlKey('f')) {
                self.selectClear();
                try self.find();
            } else if (c == ctrlKey('r')) {
                self.selectClear();
                try self.findReplace();
            } else if (c == ctrlKey('g')) {
                self.selectClear();
                try self.gotoLine();
            } else if (c == ctrlKey('c')) {
                self.copy();
                self.selectClear();
            } else if (c == ctrlKey('x')) {
                try self.cut();
            } else if (c == ctrlKey('v')) {
                try self.paste();
            } else if (c == '\r') {
                if (self.selection != null) {
                    try self.deleteSelection();
                }
                try self.insertNewline();
            } else if (c == 127 or c == ctrlKey('h')) {
                // Backspace
                if (self.selection != null) {
                    try self.deleteSelection();
                } else {
                    try self.deleteChar();
                }
            } else if (c == '\t') {
                if (self.selection != null) try self.deleteSelection();
                if (self.config.expand_tabs) {
                    const spaces = self.config.tab_size - (self.cx % self.config.tab_size);
                    for (0..spaces) |_| try self.insertChar(' ');
                } else {
                    try self.insertChar('\t');
                }
            } else if (c == '\x1b' or c == ctrlKey('l')) {
                self.selectClear();
            } else if (c >= 32 and c < 127) {
                if (self.selection != null) try self.deleteSelection();
                try self.insertChar(c);
            }
        },

        .delete => {
            if (self.selection != null) {
                try self.deleteSelection();
            } else {
                try self.deleteForward();
            }
        },

        .shift_arrow_up => {
            self.selectStart();
            self.moveCursor(Editor.Key.arrow_up);
        },
        .shift_arrow_down => {
            self.selectStart();
            self.moveCursor(Editor.Key.arrow_down);
        },
        .shift_arrow_left => {
            self.selectStart();
            self.moveCursor(Editor.Key.arrow_left);
        },
        .shift_arrow_right => {
            self.selectStart();
            self.moveCursor(Editor.Key.arrow_right);
        },

        .arrow_up, .arrow_down, .arrow_left, .arrow_right => {
            self.selectClear();
            self.moveCursor(key);
        },

        .page_up => {
            self.selectClear();
            self.cy = self.row_offset;
            var t = self.screen_rows;
            while (t > 0) : (t -= 1) {
                self.moveCursor(Editor.Key.arrow_up);
            }
        },
        .page_down => {
            self.selectClear();
            self.cy = @min(self.row_offset + self.screen_rows - 1, if (self.rows.items.len > 0) self.rows.items.len - 1 else 0);
            var t = self.screen_rows;
            while (t > 0) : (t -= 1) {
                self.moveCursor(Editor.Key.arrow_down);
            }
        },
        .home => {
            self.selectClear();
            self.cx = 0;
        },
        .end => {
            self.selectClear();
            if (self.cy < self.rows.items.len) {
                self.cx = self.rows.items[self.cy].size;
            }
        },
    }

    self.quit_times = 2;
    return true;
}

/// Prompt the user for input in the message bar. Returns owned slice or null on ESC.
pub fn prompt(self: *Editor, comptime fmt: []const u8, args: anytype) !?[]u8 {
    _ = args;

    var buf = std.ArrayList(u8).empty;
    defer buf.deinit(self.allocator);

    while (true) {
        // Format status message with current input
        var msg_buf: [80]u8 = undefined;
        const msg = std.fmt.bufPrint(&msg_buf, fmt, .{buf.items}) catch buf.items;
        @memcpy(self.status_msg[0..msg.len], msg);
        self.status_msg_len = msg.len;
        self.status_time = std.time.timestamp();
        try self.refreshScreen();

        const key = try readKey();
        switch (key) {
            .char => |c| {
                if (c == '\x1b') {
                    self.setStatusMessage("", .{});
                    return null;
                } else if (c == '\r') {
                    if (buf.items.len > 0) {
                        self.setStatusMessage("", .{});
                        return try self.allocator.dupe(u8, buf.items);
                    }
                } else if (c == 127 or c == ctrlKey('h')) {
                    if (buf.items.len > 0) {
                        buf.items.len -= 1;
                    }
                } else if (c >= 32 and c < 127) {
                    try buf.append(self.allocator, c);
                }
            },
            .delete => {
                if (buf.items.len > 0) {
                    buf.items.len -= 1;
                }
            },
            else => {},
        }
    }
}

File: src/editor/input.c
#include "editor.h"

int editor_read_key(void) {
    int nread;
    char c;
    while ((nread = read(STDIN_FILENO, &c, 1)) != 1) {
        if (nread == -1 && errno != EAGAIN) die("read");
    }

    if (c == '\x1b') {
        char seq[5];
        if (read(STDIN_FILENO, &seq[0], 1) != 1) return '\x1b';
        if (read(STDIN_FILENO, &seq[1], 1) != 1) return '\x1b';

        if (seq[0] == '[') {
            if (seq[1] >= '0' && seq[1] <= '9') {
                if (read(STDIN_FILENO, &seq[2], 1) != 1) return '\x1b';
                if (seq[2] == '~') {
                    switch (seq[1]) {
                        case '1': return HOME_KEY;
                        case '3': return DEL_KEY;
                        case '4': return END_KEY;
                        case '5': return PAGE_UP;
                        case '6': return PAGE_DOWN;
                        case '7': return HOME_KEY;
                        case '8': return END_KEY;
                    }
                }
                if (seq[1] == '1' && seq[2] == ';') {
                    char mod, dir;
                    if (read(STDIN_FILENO, &mod, 1) != 1) return '\x1b';
                    if (read(STDIN_FILENO, &dir, 1) != 1) return '\x1b';
                    if (mod == '2') {
                        switch (dir) {
                            case 'A': return SHIFT_ARROW_UP;
                            case 'B': return SHIFT_ARROW_DOWN;
                            case 'C': return SHIFT_ARROW_RIGHT;
                            case 'D': return SHIFT_ARROW_LEFT;
                        }
                    }
                }
            } else {
                switch (seq[1]) {
                    case 'A': return ARROW_UP;
                    case 'B': return ARROW_DOWN;
                    case 'C': return ARROW_RIGHT;
                    case 'D': return ARROW_LEFT;
                    case 'H': return HOME_KEY;
                    case 'F': return END_KEY;
                }
            }
        } else if (seq[0] == 'O') {
            switch (seq[1]) {
                case 'H': return HOME_KEY;
                case 'F': return END_KEY;
            }
        }

        return '\x1b';
    }

    return c;
}

char *editor_prompt(char *prompt, void (*callback)(char *, int)) {
    size_t bufsize = 128;
    char *buf = malloc(bufsize);
    size_t buflen = 0;
    buf[0] = '\0';

    while (1) {
        editor_set_status_message(prompt, buf);
        editor_refresh_screen();

        int c = editor_read_key();
        if (c == DEL_KEY || c == CTRL_KEY('h') || c == BACKSPACE) {
            if (buflen != 0) buf[--buflen] = '\0';
        } else if (c == '\x1b') {
            editor_set_status_message("");
            if (callback) callback(buf, c);
            free(buf);
            return NULL;
        } else if (c == '\r') {
            if (buflen != 0) {
                editor_set_status_message("");
                if (callback) callback(buf, c);
                return buf;
            }
        } else if (!iscntrl(c) && c < 128) {
            if (buflen == bufsize - 1) {
                bufsize *= 2;
                buf = realloc(buf, bufsize);
            }
            buf[buflen++] = c;
            buf[buflen] = '\0';
        }

        if (callback) callback(buf, c);
    }
}

void editor_process_keypress(void) {
    static int quit_times = 2;
    int c = editor_read_key();

    switch (c) {
        case '\r':
            if (editor.selecting) editor_delete_selection();
            else editor_insert_newline();
            break;

        case CTRL_KEY('q'):
            if (editor.dirty && quit_times > 0) {
                editor_set_status_message(
                    "WARNING: Unsaved changes. Press Ctrl-Q %d more time(s) to quit.",
                    quit_times);
                quit_times--;
                return;
            }
            clear_screen();
            exit(0);
            break;

        case CTRL_KEY('s'):
            editor_save();
            break;

        case CTRL_KEY('f'):
            editor_select_clear();
            editor_find();
            break;

        case CTRL_KEY('r'):
            editor_select_clear();
            editor_find_replace();
            break;

        case CTRL_KEY('o'):
            editor_select_clear();
            editor_switch_file();
            break;

        case CTRL_KEY('g'):
            editor_select_clear();
            editor_goto_line();
            break;

        case CTRL_KEY('c'):
            editor_copy();
            editor_select_clear();
            break;

        case CTRL_KEY('x'):
            editor_cut();
            break;

        case CTRL_KEY('v'):
            editor_paste();
            break;

        case HOME_KEY:
            editor_select_clear();
            editor.cpos.x = 0;
            break;
        case END_KEY:
            editor_select_clear();
            if (editor.cpos.y < editor.rows)
                editor.cpos.x = editor.row[editor.cpos.y].size;
            break;

        case BACKSPACE:
        case CTRL_KEY('h'):
            if (editor.selecting) editor_delete_selection();
            else editor_del_char();
            break;
        case DEL_KEY:
            if (editor.selecting) {
                editor_delete_selection();
            } else {
                editor_move_cursor(ARROW_RIGHT);
                editor_del_char();
            }
            break;

        case SHIFT_ARROW_UP:
        case SHIFT_ARROW_DOWN:
        case SHIFT_ARROW_LEFT:
        case SHIFT_ARROW_RIGHT:
            editor_select_start();
            switch (c) {
                case SHIFT_ARROW_UP:    editor_move_cursor(ARROW_UP);    break;
                case SHIFT_ARROW_DOWN:  editor_move_cursor(ARROW_DOWN);  break;
                case SHIFT_ARROW_LEFT:  editor_move_cursor(ARROW_LEFT);  break;
                case SHIFT_ARROW_RIGHT: editor_move_cursor(ARROW_RIGHT); break;
            }
            break;

        case ARROW_UP:
        case ARROW_RIGHT:
        case ARROW_DOWN:
        case ARROW_LEFT:
            editor_select_clear();
            editor_move_cursor(c);
            break;

        case PAGE_UP:
        case PAGE_DOWN: {
            editor_select_clear();
            if (c == PAGE_UP) {
                editor.cpos.y = editor.row_offset;
            } else {
                editor.cpos.y = editor.row_offset + editor.win_size.rows - 1;
                if (editor.cpos.y > editor.rows) editor.cpos.y = editor.rows;
            }
            int t = editor.win_size.rows;
            while (t--) editor_move_cursor(c == PAGE_UP ? ARROW_UP : ARROW_DOWN);
            break;
        }

        case CTRL_KEY('l'):
        case '\x1b':
            editor_select_clear();
            break;

        case '\t':
            if (editor.selecting) editor_delete_selection();
            if (config.expand_tabs) {
                int spaces = config.tab_size - (editor.cpos.x % config.tab_size);
                for (int i = 0; i < spaces; i++) editor_insert_char(' ');
            } else {
                editor_insert_char('\t');
            }
            break;

        default:
            if (c >= 32 && c < 127) {
                if (editor.selecting) editor_delete_selection();
                editor_insert_char(c);
            }
            break;
    }

    quit_times = 2;
}
